calcom/cal.com
# Beginning of Strict CSP Compliance

## What does this PR do?

**CSP Policy won't be added automatically and it is Opt In using CSP_POLICY env variable**

- Added 2 more security headers apart from Content-Security-Policy
	- X-Content-Type-Options
	- Referrer-Policy
- Enforces Strict CSP for the login page
	- If the user goes to any other route without a full page refresh, that would also follow strict CSP Policy restrictions. 
- Enables CSP in Report Only mode for all SSR Pages. So, nothing breaks there. It would just report the errors that come. Wherever a component is adding a script tag it would have to use `withNonce`.
- Introduces env variable CSP_POLICY with 2 possible values `strict`, `non-strict`
- Added a response header `x-nonce` with values `"ssr" | "initialPropsOnly"|"false"|"not-opted-in"`
	- `ssr` -> Page is an SSR Page and Component will not have `nonce` available as prop
	- `initialPropsOnly` ->  Page is probably an SSR Page but Components won't have `nonce` available as prop
	- `false` -> Due to some reason nonce wasn't available and thus CSP is disabled for the page.
	- `not-opted-in` -> CSP could be enabled but CSP_POLICY env variable isn't set.
	- If the header is not present, then CSP is not enabled for the page(e.g. SSG pages)

_Parser Inserted Script(without nonce) is blocked and img src that's not whitelisted is also blocked_
<img width="1792" alt="Screenshot 2023-02-02 at 6 38 52 PM" src="https://user-images.githubusercontent.com/1780212/216333576-ac34b6e3-d4e9-4409-87b6-5ddc98535478.png">

_CSP Evaluator Status for the Policy_
<img width="1093" alt="Screenshot 2023-02-02 at 6 51 50 PM" src="https://user-images.githubusercontent.com/1780212/216336301-7b4b375d-4c1c-4db4-a49f-2bb93028e52c.png">

_securityheaders.com status for /auth/login_
<img width="1105" alt="Screenshot 2023-02-02 at 7 52 26 PM" src="https://user-images.githubusercontent.com/1780212/216350344-22c222ba-3a6f-4057-9c9b-2561841d5f24.png">

# Important
**CSP_POLICY='non-strict' should be set to enable CSP in production. At any moment if we see any issues due to CSP, we need to simply remove this env variable and redeploy.**

## Followup
- [ ] Add a report-to directive that would log all CSP reports to Sentry.
- [ ] Find a way to fix [this error](https://reactjs.org/docs/error-decoder.html/?invariant=421) that comes only sometimes.
- [ ] Enforce CSP for other pages which has Report Only enabled.
- [ ] Figure out how to do it for SSG and ISR
- [ ] Add Permissions-Policy - We would probably be restricting all features like geolocation, camera access
- [ ] Explore require-trusted-types for directive
- [ ] Publish findings from https://www.notion.so/hariombalhara/CSP-Implementation-Cal-com-a9580b937fd441008dd4d81396bdc03b onto Github or gitbook under various sections like
	-  **Debugging CSP issues**
	- **Differences in CSP b/w local and production modes** 
	- **How to test CSP**



##
**Environment**:  Production

## Type of change




## How should this be tested?

Please ensure that you test the app in an instance with no browser extensions(usually incognito mode doesn't have browser extensions so that can be the quickest way to start testing). **This is important to avoid false failures arising due to resources requested by one of your extensions**

- [ ] Test that on Safari(<=15.3), Safari latest, Chrome and Firefox login page is working
- [ ] Also test that after a fresh login, when you browse the app all the pages are functional and all images are also loading. This is important because browsing after a fresh login keeps the entire app under CSP restrictions as all navigations happen in SPA flow. After refresh on any non login page, CSP won't apply for now.

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works
